bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AutoRecoveryMain.java

```diff
@@ -91,6 +92,11 @@ public AutoRecoveryMain(ServerConfiguration conf, StatsLogger statsLogger)
             CompatibilityException {
         this.conf = conf;
         this.bkc = Auditor.createBookKeeperClient(conf);
+        MetadataClientDriver metadataClientDriver = bkc.getMetadataClientDriver();
+        metadataClientDriver.setConnectionExpiredListener(() -> {
```

Comment:
I think a cleaner solution is AuditorElector handling this session expires. because AuditorElector already has this session expires watcher to shutdown elector - https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java#L224 so all the leader election logic are all in one place, rather than spreading across multiple places. and it seems to be working better with #1701 (https://github.com/apache/bookkeeper/pull/1701/files#diff-7d01af204bb0a30c0193b46c737f8cbfR49)

However I am fine with this change as it is working well with the zookeeper concepts. We can go with this change for now, once the change #1701 is merged, we can considering moving this session expires logic in to the AuditorElector, which will provide a much cleaner abstraction.

Reply:
@sijie thank you for pointing this out. yes, I think session expires is an implementation specific behavior, I will incorporate your proposal into #1701 once this change is merged.